### PR TITLE
Refactor intl to use next context

### PR DESCRIPTION
**Pull request by hdiniz**
Resolve https://github.com/opencollective/opencollective/issues/6795

# Description

Removes existing intl middleware in favor of using next context to resolve the user locale and load messages.

Before

- All i18n messages are bundled on the SSR html, messages are included on every full page load from the server.

image

After

- SSR html contains only the messages used in the render.
- Hydratation and rest of i18 messages are delivered on a separate chunk.
- Messages can be cached independently of the html page.
- switching languages without full page reload is possible.

image

language-switch

### CODE REVIEW

File: lib/i18n/request.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    locale,
+    messages,
+  };
+}
+
+export function getRequestIntl(req: NextPageContext['req']): Pick<IntlProps, 'language' | 'locale'> {
+  let language: string, locale: string;
+
+  const url = new URL(req.url, `http://${req.headers.host}`);
+  const queryLanguage = url.searchParams.get('language');
```
Comment (Betree) [inline lib/i18n/request.ts:41]: I don't get this bit: isn't the host supposed to contain only the domain, e.g. `opencollective.com`, without any URL parameters?

File: pages/_document.js [hunk 2]
```diff
@@ -95,12 +91,6 @@ export default class IntlDocument extends Document {
         <Head nonce={this.props.cspNonce} />
         <body>
           <Main nonce={this.props.cspNonce} />
-          <script
-            nonce={this.props.cspNonce}
-            dangerouslySetInnerHTML={{
-              __html: this.props.localeDataScript,
-            }}
-          />
```
Comment (Betree) [inline pages/_document.js:103]: Good to see that thing go away!

**Review by Betree — COMMENTED** [review]
The code looks very good so far! A few questions on things that are not clear to me below.

Also, on the deployment preview, it doesn't work when I try changing the language in the footer. Is it supposed to be supported for Vercel?

File: lib/i18n/request.ts (hunk 1 above)
Comment (hdiniz) [reply lib/i18n/request.ts:41]: `req.url` contains the request path + params 

js docs:
```
Request URL string. This contains only the URL that is present in the actual HTTP request. Take the following request:

GET /status?name=ryan HTTP/1.1
Accept: text/plain
To parse the URL into its parts:

new URL(request.url, `http://${request.headers.host}`);
When request.url is '/status?name=ryan' and request.headers.host is'localhost:3000':

$ node
> new URL(request.url, `http://${request.headers.host}`)
URL {
  href: 'http://localhost:3000/status?name=ryan',
  origin: 'http://localhost:3000',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'localhost:3000',
  hostname: 'localhost',
  port: '3000',
  pathname: '/status',
  search: '?name=ryan',
  searchParams: URLSearchParams { 'name' => 'ryan' },
  hash: ''
}
```

**Comment by hdiniz** [comment]
@Betree there is a caching issue, the `en` version is being cached on vercel. It works when there is a cache miss. I though this was resolved, I will have another look.

File: lib/i18n/request.ts (hunk 1 above)
Comment (Betree) [reply lib/i18n/request.ts:41]: Interesting, thanks for sharing

**Review by znarf — APPROVED** [review]

**Comment by znarf** [comment]
Great work! Let's try that

**Pull request merged**